Use a property to enable OpenTelemetry for JDBC and use the OpenTelemetry datasource instead of the OpenTelemetry driver

michalvavrik opened the pull request:
closes: #30532

Allow user to enable JDBC tracing by wrapping `AgroalDataSource` with the `OpenTelemetryDatasource` when the configuration property `quarkus.datasource.jdbc.tracing=true` is set. Usage of the OpenTelemetry driver is no longer supported (we need this change because of SDK autoconfiguration).

gsmet commented on docs/src/main/asciidoc/opentelemetry.adoc:
```diff
@@ -243,15 +243,16 @@ The https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/ma
 ----
 
-As it uses a dedicated JDBC driver, you must configure your datasource and Hibernate ORM to use it.
+As it uses a dedicated JDBC driver, you must enable JDBC tracing for your datasource and Hibernate ORM to use it.
```
```suggestion
As it uses a dedicated JDBC driver, you must enable JDBC tracing for your datasource:
```

gsmet commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -205,6 +214,10 @@ void generateDataSourceSupportBean(AgroalRecorder recorder,
             BuildProducer<UnremovableBeanBuildItem> unremovableBeans) {
         additionalBeans.produce(new AdditionalBeanBuildItem(JdbcDriver.class));
 
+        if (capabilities.isPresent(Capability.OPENTELEMETRY_TRACER)) {
+            recorder.useOpenTelemetryForTracing();
+        }
```
I'm not sure I understand why you need that? And why you didn't follow the pattern we used for OpenTracing. I'm probably missing something but I would like to understand.

gsmet replied:
Let's go with `.telemetry` and have a different setting. Then you can be consistent and use the same pattern.

If one day, we remove OpenTracing support, it will also be easier.

gsmet commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSourceJdbcTracingRuntimeConfig.java:
```diff
@@ -15,13 +15,15 @@ public class DataSourceJdbcTracingRuntimeConfig {
     /**
-     * Trace calls with active Spans only
+     * Trace calls with active Spans only.
+     * Please note that this configuration property is only supported with OpenTracing.
```
I think the plan is to retire OpenTracing so I would rather not add this line. But @radcortez will confirm.

michalvavrik replied:
@gsmet this property can't be there for OpenTelemetry as it is not used by it. If OpenTracing is going to be retired, we can remove the property. There need to be a way to tell users to ignore it for OTEL. @radcortez WDYT?

brunobat replied:
OpenTracing will not be retired immediately. We haven't decided on sunset process.

gsmet commented on integration-tests/opentelemetry-jdbc-instrumentation/src/main/resources/application.properties:
```diff
@@ -7,37 +7,40 @@ quarkus.datasource.devservices.enabled=false
 
 # JDBC instrumentation setting
-driver=io.opentelemetry.instrumentation.jdbc.OpenTelemetryDriver
+otel-driver=io.opentelemetry.instrumentation.jdbc.OpenTelemetryDriver
```
Why do you need that? Wouldn't you just need to enable tracing?

michalvavrik replied:
I wanted to make sure previous behavior still works. There are 2 datasources using tracing flag and 2 that does not. It verifies backwards compatibility.

gsmet reviewed (changes requested):
Thanks for that. It's something that has been in the way for a long time.

I added a few questions/comments.

I'm also not sure if we want to reuse the `tracing` configuration property. I think the idea is to retire OpenTracing at some point so maybe having a separate `telemetry` boolean for telemetry would be more future proof and less confusing (we would use the same term everywhere).

I will let @brunobat and @radcortez express their opinions on this!

brunobat commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -97,14 +97,23 @@ void build(
+                    tracingDriver = DataSources.TRACING_DRIVER_CLASSNAME;
+                } else {
+                    throw new ConfigurationException(
+                            "JDBC instrumentation has been enabled with configuration property 'quarkus.datasource.jdbc.tracing', but neither Quarkus OpenTelemetry nor SmallRye OpenTracing are present");
```
I think we need tests on this logic ate the extension level and not only on IT tests.

michalvavrik replied:
That logic will be gone when we go with datasource way, so I didn't add unit tests for it.

michalvavrik commented:
Please note that scope of this PR has changed. Now OpenTelemetry JDBC instrumentation uses OTEL datasource wrapper as SDK autoconfiguration will require runtime init of OTEL and that's not compatible with OpenTelemetryDriver. The OTEL driver is no longer supported (breaking change).

michalvavrik commented:
Hmm, I tried non-telemetry JPA modules, but didn't think of native. I'll fix failures.

brunobat commented:
Moving to the OTel datasource way is required for the upcoming https://github.com/quarkusio/quarkus/pull/30033

michalvavrik commented:
I moved binding between Agroal and OpenTelemetry to deployment level (byte code only generated when the telemetry is active). Alas, fixed the problem, but looks less elegant.

brunobat commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/OpenTelemetryJDBCInstrumentationActiveBuildItem.java:
```diff
@@ -0,0 +1,9 @@
+/**
+ * Purely marker build item that tells us to prepare OpenTelemetry JDBC instrumentation.
+ */
+final class OpenTelemetryJDBCInstrumentationActiveBuildItem extends SimpleBuildItem {
```
Just `OpenTelemetryJDBCInstrumentationBuildItem`?

brunobat commented on docs/src/main/asciidoc/opentelemetry.adoc:
```diff
@@ -243,15 +243,16 @@ The https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/ma
 ----
 
-As it uses a dedicated JDBC driver, you must configure your datasource and Hibernate ORM to use it.
+As it uses a dedicated JDBC driver, you must enable JDBC tracing for your datasource:
```
As it uses a dedicated Datasource wrapper, you must enable JDBC tracing for your datasource:

michalvavrik replied:
Thanks. Now I also mentioned `JDBC driver`, that it not true anymore.

brunobat commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -105,13 +108,23 @@ void build(
             aggregatedConfig.produce(aggregatedDataSourceBuildTimeConfig);
         }
 
+        if (otelJdbcInstrumentationActive) {
```
This should only be active if the OTel capability is available.

brunobat commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/OpenTelemetryAgroalDataSource.java:
```diff
@@ -0,0 +1,66 @@
+    private final AgroalDataSource delegate;
+
+    public OpenTelemetryAgroalDataSource(AgroalDataSource delegate) {
+        super(delegate, GlobalOpenTelemetry.get());
```
We must not use GlobalOpenTelemetry.get().
I wonder if we don't need to send this `AgroalDataSource` to the OTel extension and wire `OpenTelemetryAgroalDataSource` there.
We need to have a way to set the OpenTelemetry object later.

michalvavrik replied:
I realize that, but I thought it would be the easiest to refactor with your changes in #30033. Anyway, point taken, I'll read up #30033 and prepare solution that will not require changes later.

michalvavrik replied:
I'll introduce OT SPI and move things around.

brunobat replied:
What you suggested works for me. No worries.

michalvavrik replied:
Thanks @brunobat , right now we can't really force order that first, OpenTelemetry is created and then step that set up datasources and uses OTEL runs because (Tracer ->) OpenTelemetry must run after CDI bean container is fully initialized, while Agroal step produces additional bean [circular reference].

I think best solution would be to refactor it based on your PR (now everything works as OTEL is initialized during static init). I'm happy to do that, just not here as you would hit merge hell.

brunobat reviewed (approved):
Thanks @michalvavrik. Much Appreciated.

gsmet committed:
Use a bean for telemetry wrapping to avoid making DataSources mutable

We had a hard time making it immutable so let's not break this contract.
We had a lot of race conditions before doing so.

gsmet commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -111,7 +110,7 @@ void build(
             }
 
-            if (aggregatedDataSourceBuildTimeConfig.getJdbcConfig().telemetry && !otelJdbcInstrumentationActive) {
+            if (aggregatedDataSourceBuildTimeConfig.getJdbcConfig().telemetry) {
```
I simplified the `if` here as the extra condition was making it a bit more obscure than truly necessary.

michalvavrik replied:
My thinking was that boolean evaluation is quicker than variable assignment :-) Simplified `if` is fine.

gsmet replied:
So yes in theory. It might be of importance in a very hot path but it's not one and it's probably better to keep the code more readable.
But as I said, this one is debatable. Your pattern is more efficient but my impression is that it isn't worth reducing the readability.

gsmet commented on extensions/agroal/deployment/src/main/java/io/quarkus/agroal/deployment/AgroalProcessor.java:
```diff
@@ -124,8 +123,10 @@ void build(
+            // therefore we register the OpenTelemetry data source wrapper bean
+            additionalBeans.produce(new AdditionalBeanBuildItem.Builder()
+                    .addBeanClass("io.quarkus.agroal.runtime.AgroalOpenTelemetryWrapper")
+                    .setDefaultScope(DotNames.SINGLETON).build());
```
We now produce a bean that we will be able to inject into `DataSources` without making it mutable.

michalvavrik replied:
Sure, that's better.

gsmet commented on extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/DataSources.java:
```diff
@@ -254,10 +257,11 @@ public AgroalDataSource doCreateDataSource(String dataSourceName) {
         }
 
-        if (dataSourceJdbcBuildTimeConfig.telemetry && dataSourceJdbcRuntimeConfig.telemetry.orElse(true)) {
+        if (dataSourceJdbcRuntimeConfig.telemetry.orElse(dataSourceJdbcBuildTimeConfig.telemetry)
```
I think using this pattern makes it the intent a bit clearer.

michalvavrik replied:
@gsmet This condition will pass when OT is enabled at runtime, but disabled at build time. There are steps that need to be done at build time (like creating additional bean you produced)

gsmet replied:
I'm not sure I understand your comment.

gsmet replied:
Thanks for pushing back, I will push an update in a few minutes.

michalvavrik replied:
Thanks, please consider also moving the DEV mode test as then you can remove the hibernate dependency @gsmet

gsmet replied:
I'm not sure I understand where you want to move the test. Could you take it from there?

michalvavrik replied:
Sure, will do.

gsmet reviewed (commented):
I added an additional commit (and rebased in passing). See my explanations inline.
The important change is about avoiding making `DataSources` mutable (again!) as we worked hard to avoid that.
The other changes are more debatable as they are about personal preferences but I thought they could be interesting to you anyway.

There's one change left to be done (providing you agree with my changes, of course), it's avoiding to add the Hibernate ORM test dependency (except if strongly necessary, tell me if it's the case).

Also, I'm not sure what you had in mind for the future and avoiding the use of the global OpenTelemetry thing so let me know if what I did causes issues.

michalvavrik commented:
This pattern is better and it won't block as later regarding obtaining the OpenTelemetry. Thanks

brunobat commented:
I really like this new approach and seems to me that will make it easier the integration with the upcoming otel runtime instantiation.

michalvavrik committed:
Use a property to enable OTEL for JDBC & replace driver with datasource

michalvavrik commented on extensions/opentelemetry/deployment/pom.xml:
```diff
@@ -116,11 +116,22 @@
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-orm-deployment</artifactId>
+            <scope>test</scope>
+        </dependency>
```
The dependency is gone.

michalvavrik commented:
Is this one fine to go? I'd like to avoid conflicts with the other OT PR from Bruno.

gsmet reviewed (approved):
Yes, let's merge it. Nice team work!

famod commented:
Is it just me or is this change missing from the 3.0 migration guide?

brunobat commented:
You are right @famod. Fixed now.
